### ISSUE TITLE
How to skip records in the CSV?

### ISSUE THREAD

**Issue by kapgit** [awdeorio/mailmerge#85]
### The context: Using a DB as primary datasource

I'm using mailmerge, using a MySQL database as datasource.
This DB contains data and a number of custom views. Each view queries the data and returns a row for each mail to be sent with mailmerge.

As mailmerge is accepting only CSV file as input (as far as I know), and not a DB view, a little utility bash script 
```bash
send_to <view> <mail_template>
```
is executing the DB \<view\> and generates a transient CSV file containing the selected rows.

The script then calls mailmerge using the path to this CSV file and a Jinja2 \<mail_template\> as input parameters.

Mailmerge does its job and sends a mail for each line in the CSV file. Perfect!

### Issue:

Sometimes, I need to quickly send a mail to a subset of the mysql \<view\>. One shot. No need to reuse the query.
I could of course create/alter the DB view, but that's not as agile as I would like and could quickly lead to proliferation of views, all with their own little variations on some 'master' view.

### Question:

Is is possible to use some Jinja2 tagging in the mail template, that would have as an effect to skip the current record?

Something like:

```
{% if condition_based_on_current_record -%}
TO: {{email}}
SUBJECT: Interesting mail
FROM: My Self <myself@mydomain.com>

Dear {{name}},

You are the lucky recipient of this email.

Best regards,

Me
{% endif -%}
```

If condition_based_on_current_record is true then the mail is sent.
Otherwise, the sender, recipient and message are all empty, and mailmerge would skip the record.

Thank you.

**Comment by kapgit** [comment]
Maybe there is a better method, but at the moment, I'm applying a quick and dirty solution that seems to work for me:

In template_message.py (commented the 3 assertions):
```python
    def render(self, context):
        """Return rendered message object."""
        try:
            raw_message = self.template.render(context)
        except jinja2.exceptions.TemplateError as err:
            raise MailmergeError(
                "{}: {}".format(self.template_path, err)
            )
        self._message = email.message_from_string(raw_message)
        self._transform_encoding(raw_message)
        self._transform_recipients()
        self._transform_markdown()
        self._transform_attachments()
        self._message.__setitem__('Date', email.utils.formatdate())
        #assert self._sender # COMMENTED
        #assert self._recipients # COMMENTED
        #assert self._message # COMMENTED
        return self._sender, self._recipients, self._message
```

In \_\_main\_\_.py (enclosed the 'sendmail' in an 'if' statement):
```python
def main(sample, dry_run, limit, no_limit, resume,
         template_path, database_path, config_path,
         output_format):
...
   try:
        template_message = TemplateMessage(template_path)
        csv_database = read_csv_database(database_path)
        sendmail_client = SendmailClient(config_path, dry_run)
        for _, row in enumerate_range(csv_database, start, stop):
            sender, recipients, message = template_message.render(row)
            if sender and recipients and message: # ADD THIS LINE + INDENT FOLLOWING
                sendmail_client.sendmail(sender, recipients, message)
                print_bright_white_on_cyan(
                    ">>> message {message_num}"
                    .format(message_num=message_num),
                    output_format,
                )
                print_message(message, output_format)
                print_bright_white_on_cyan(
                    ">>> message {message_num} sent"
                    .format(message_num=message_num),
                    output_format,
                )
                message_num += 1

    except MailmergeError as error:
...
```

**Comment by awdeorio** [comment]
One way to do it is to make a copy of the CSV file and delete some of the rows with a text editor.  Another way to do it is to filter the CSV file using command line tools like `grep` or [`csvkit`](https://csvkit.readthedocs.io/en/latest/).

Finally, here's an example of how to use the command line options to skip a message, e.g., message 11.
```console
$ mailmerge --limit 10
$ mailmerge --resume 12
```

**Comment by bexelbie** [comment]
It'd be great to see a flag or similar for this if we can do it in a UI friendly way.  I have a routine email that I send that needs to come from one of two email addresses.  Today I have to bifurcate my data set and reset the conf for each run.  Not having to bifurcate the data set would be a small win.

**Comment by awdeorio** [comment]
Here's a brain dump of some different ways this could work.  Feedback welcome.

1. Include row numbers, similar to the `cut -f` option.  From `man cut`:
    ```console
    $ mailmerge --include-rows 1-10,13-20
    $ man cut
    ...
    Use one, and only one of -b, -c or -f.  Each LIST is  made  up  of  one
    range,  or  many ranges separated by commas.  Selected input is written
    in the same order that it is read, and is written exactly  once.   Each
    range is one of:
    
    N      N'th byte, character or field, counted from 1
    
    N-     from N'th byte, character or field, to end of line
    
    N-M    from N'th to M'th (included) byte, character or field
    
    -M     from first to M'th (included) byte, character or field
    ```

2. Exclude row numbers, e.g.:
    ```console
    $ mailmerge --exclude-rows 11-12
    ```

3. Pattern match against a column, similar to [`csvgrep`](https://csvkit.readthedocs.io/en/1.0.2/scripts/csvgrep.html).
    ```console
    $ mailmerge --column email --match umich.edu
    $ mailmerge --column email --match umich.edu --invert-match
    ```

4. Skip message if entire rendered template is blank.  This is similar to the suggestion of @kapgit above.  I'd perhaps couple it with a command line flag.  Perhaps check for blank messages and produce a helpful error message, suggesting this flag.
    ```console
    $ mailmerge --skip-blank
    ```

    And the corresponding `mailmerge_template.txt`:
    ```
    {% if condition_based_on_current_record -%}
    TO: {{email}}
    FROM: from@test.com

    Hello world
    {% endif -%}
    ```

5. Skip message if `TO` field is blank.
    ```console
    $ mailmerge --skip-blank-to
    ```

    And the corresponding `mailmerge_template.txt`:
    ```
    {% if condition_based_on_current_record -%}
    TO: {{email}}
    {% endif -%}

    FROM: from@test.com

    Hello world
    ```

I'm leaning towards option 4, similar to @kapgit 's suggestion above.

**Comment by bexelbie** [comment]
@awdeorio I like what you have. FWIW, I'd like to express my preference for number 5 because it is smaller and keeps the if small to prevent weird nesting typos later.

**Comment by captn3m0** [comment]
+1 for (5). Empty to currently results in a assertion error, so its anyway not a supported usecase.

**Comment by awdeorio** [comment]
I can get behind (5).  Any volunteers? :)

**Comment by bzoel** [comment]
I'd like to take a swing at this one (just a few years later).

One clarification regarding the CC, BCC behavior discussed above. Options:

1. Behave like `--skip-blank-recipients`
    Message skipped if TO, CC, and BCC are all blank
    Message sent if TO is blank, but CC or BCC are not blank

2. Behave like `--skip-blank-to`
    Message skipped if TO is blank without considering CC and BCC

I was leaning toward (1), because it also allows this to be used for other use cases, i.e. sending a BCC only message when it is intended to hide the recipients list.

**Comment by UffeJakobsen** [comment]
Agree - (1) is the most flexible

**Comment by awdeorio** [comment]
Agree, skip if there are no recipients.  One more thought.  I imagine that this isn't something the user would change frequently for a given group of messages.  Would it be better to put the `skip-blank-recipients` in `mailmerge_server.conf` ?

**Comment by bzoel** [comment]
Good point there, I imagine this is more of a "set it and forget it". Maybe something like this in `mailmerge_server.conf` ?

`
[advanced]
skip_blank_recipients = True
`

**Comment by awdeorio** [comment]
> Does it seem make more sense to move the whole read_config functionality to `main.py` with this change?

Yes, I agree that makes sense.

What do you think about something like `[template]` or `[template_options]` instead of `[advanced]`?
